README.md
# Facemovie

This project aims to produce videos in the style of "picasa face movie".
Example (generated by picasa) here: https://www.youtube.com/watch?v=fLQtssJDMMc

## Usage

1. Create a python environment (or don't, you do you)

    python -mvenv env
    . env/bin/activate
   
2. Install dependencies

    pip install -r requirements.txt
   
3. Run the gui to select images and faces. When you're done, save your project. It's a json file but the extension doesn't matter.

    python gui.py

## CLI usage

When you have a project file, you can generate the movie from the command line.

    python generate_movie.py <project.json> movie.mov

## possible improvements:

### functionality
* Work around the possible mismatches between faces detected by `mediapipe` and `face_recognition` libraries
* GUI: allow adding single image
* GUI: allow removing single image
* GUI: allow reordering images

### ux improvement
* Action buttons with icons
* menu bar

### Movie improvements
* Sometimes the oriented image eyes are not exactly horizontal. Debug and fix this
* Allow picking the eye target location

### various
* Packaging
* Choose a non-generic file extension for project files
* Improve CLI: allow overwriting settings from project json


CHANGELOG.md
# Next release

- added borders around images
- generate facemovie from the GUI
- remember last opened project

run.py
import sys

import cv2
import mediapipe as mp
import numpy as np

LEFT_EYE_CORNER_LM = 7
RIGHT_EYE_CORNER_LM = 249
CHIN_LM = 152


def get_face_landmarks(face_mesh, img):
    for i in range(20):
        mesh_result = face_mesh.process(img)
    return mesh_result


def determine_angle(a, b):
    unit_a = a / np.linalg.norm(a)
    unit_b = b / np.linalg.norm(b)
    dot_product = np.dot(unit_a, unit_b)
    return np.arccos(dot_product)


def rotate_vector(vector, angleA_rad):
    return np.array([[np.cos(angleA_rad), -np.sin(angleA_rad)], [np.sin(angleA_rad), np.cos(angleA_rad)]]).dot(vector)


def process_image(face_mesh, img):
    mesh_result = get_face_landmarks(face_mesh, img)
    left_eye_lm = mesh_result.multi_face_landmarks[0].landmark[LEFT_EYE_CORNER_LM]
    right_eye_lm = mesh_result.multi_face_landmarks[0].landmark[RIGHT_EYE_CORNER_LM]
    chin_lm = mesh_result.multi_face_landmarks[0].landmark[CHIN_LM]

    output_height = 1024
    output_width = int(1024 * 3 / 2)

    input_height = img.shape[0]
    input_width = img.shape[1]

    left_orig = np.array([input_width * left_eye_lm.x, input_height * left_eye_lm.y])
    right_orig = np.array([input_width * right_eye_lm.x, input_height * right_eye_lm.y])
    chin_orig = np.array([input_width * chin_lm.x, input_height * chin_lm.y])

    left_target = np.array([output_width * 0.47, output_height / 2])
    right_target = np.array([output_width * 0.53, output_height / 2])

    chin_target = calculate_chin_target(chin_orig, left_orig, left_target, right_orig, right_target)

    # note: it was probably easier to manually create the transformation matrix based just on the eye coordinates
    # try that later.
    M = cv2.getAffineTransform(
        np.float32([left_orig, right_orig, chin_orig]),
        np.float32([left_target, right_target, chin_target]))
    img_alpha = add_alpha_channel(img)
    overlay = cv2.warpAffine(img_alpha, M, (output_width, output_height))

    return img, overlay


def calculate_chin_target(chin_orig, left_orig, left_target, right_orig, right_target):
    AB = right_orig - left_orig
    AC = chin_orig - left_orig
    angleA = determine_angle(AB, AC)
    ab = right_target - left_target
    ab_rotated = rotate_vector(ab, angleA)
    ac = ab_rotated * np.linalg.norm(AC) / np.linalg.norm(AB)
    chin_target = left_target + ac
    return chin_target


def add_alpha_channel(img):
    channel1, channel2, channel3 = cv2.split(img)
    alpha_channel = np.ones(channel1.shape, dtype=channel1.dtype) * 255
    return cv2.merge((channel1, channel2, channel3, alpha_channel))


def load_image(num):
    img = cv2.imread('examples/me/photo%03d.jpg' % (num + 1))
    return img


def main():
    face_mesh = mp.solutions.face_mesh.FaceMesh(static_image_mode=False, max_num_faces=2)
    mp_draw = mp.solutions.drawing_utils

    current_image = 3
    current_landmark = 0
    img, result = process_image(face_mesh, load_image(current_image))

    while True:
        cv2.imshow('input', img)
        cv2.imshow('result', result)

        key = cv2.waitKey(1)
        if key == ord('a'):
            current_image += 1
            current_image %= 4
            img, result = process_image(face_mesh, load_image(current_image))
        elif key == ord('n'):
            current_landmark += 1
            current_landmark %= 500
        elif key == ord('p'):
            current_landmark += 499
            current_landmark %= 500
        elif key == ord('q'):
            sys.exit(0)


def draw_face_mesh(img, mesh_result):
    if mesh_result.multi_face_landmarks:
        # print("got landmarks", mesh_result.multi_face_landmarks)
        for face in mesh_result.multi_face_landmarks:
            mp_draw.draw_landmarks(img, face)


def draw_eye_landmarks(img, mesh_result):
    if not mesh_result.multi_face_landmarks:
        return

    landmarks = mesh_result.multi_face_landmarks[0]
    draw_landmarks(img, landmarks)


def draw_landmarks(img, landmarks):
    for i in range(len(landmarks.landmark)):
        draw_single_landmark(img, landmarks.landmark[i], str(i))


def draw_single_landmark(img, landmark, text=None):
    height = img.shape[0]
    width = img.shape[1]
    lm_pos = (int(landmark.x * width), int(landmark.y * height))
    cv2.circle(img, lm_pos, 1, (0, 255, 0), 2)
    if text is not None:
        cv2.putText(img, text, lm_pos, cv2.FONT_HERSHEY_PLAIN, 2, (255, 255, 255))


if __name__ == "__main__":
    main()


gui/main.py
import glob
import os
import sys
import threading

import cv2
import face_recognition
import numpy as np
import yaml
from PyQt5.QtCore import Qt, QStringListModel
from PyQt5.QtGui import QStandardItemModel, QStandardItem, QImage, QPixmap
from PyQt5.QtWidgets import QApplication, QMainWindow, QLabel, QFileDialog, QToolBar, QListView, \
    QHBoxLayout, QWidget, QVBoxLayout, QAbstractItemView, QAction

import project
from generate_facemovie import generate_facemovie
from gui.components.output_settings import OutputSettingsWindow
from project import Settings


def face_to_rect(face):
    return project.Rect(project.Point(face[1], face[0]), project.Point(face[3], face[2]))


class WorkQueueWindow(QWidget):
    """
    This "window" is a QWidget. If it has no parent,
    it will appear as a free-floating window.
    """

    def __init__(self, work_queue):
        super().__init__()
        self.work_queue = work_queue
        layout = QVBoxLayout()
        self.label = QLabel("Work queue")
        self.list_view = QListView()
        self.model = QStringListModel()
        self.list_view.setModel(self.model)
        layout.addWidget(self.label)
        layout.addWidget(self.list_view)
        self.setLayout(layout)

    def update(self):
        # TODO: figure out which row number to update instead of rebuilding the complete model
        items = [(item.project.settings.output_path, item.progress) for item in self.work_queue]
        self.model.setStringList(['%s - %2.2f' % item for item in items ])


class WorkItem:
    def __init__(self, project):
        self.project = project
        self.status = 'QUEUED'
        self.progress = 0.0  # percentage

    def update(self, new_status, progress=None):
        self.status = new_status
        if self.status == 'FINISHED':
            self.progress = 100
        else:
            if self.progress is not None:
                self.progress = progress


class MainWindow(QMainWindow):
    """Main Window."""

    def __init__(self, parent=None):
        """Initializer."""
        super().__init__(parent)
        self.selected_image = QLabel("selected image")
        self.image_list_model = None
        self.image_list = None

        self.work_queue = []
        self.work_queue_window = WorkQueueWindow(self.work_queue)

        self.generator_thread = None

        self._create_actions()
        self.settings = Settings()
        self.create_imagelist()

        self.current_project_path = None
        self.current_image_faces = []

        self.setWindowTitle("Facemovie builder")
        self.resize(800, 480)

        self.create_main_widget()

        toolbar = self.create_toolbar()
        self.addToolBar(Qt.TopToolBarArea, toolbar)

        rc = self.load_rcfile()
        if rc.get('last_opened', None) is not None:
            self.load_project(rc['last_opened'])
        else:
            self.start_new_project()

    def _create_actions(self):
        self.action_new = QAction("New", self)
        self.action_new.triggered.connect(self.handle_action_new)

        self.action_open = QAction("Open", self)
        self.action_open.triggered.connect(self.handle_action_open)

        self.action_save = QAction("Save", self)
        self.action_save.triggered.connect(self.handle_action_save_project)

        self.action_add_image_directory = QAction("Add images from dir", self)
        self.action_add_image_directory.triggered.connect(self.handle_action_add_image_directory)

        self.action_generate_facemovie = QAction("Generate facemovie", self)
        self.action_generate_facemovie.triggered.connect(self.handle_action_generate_facemovie)

        self.action_clear_face = QAction("clear", self)
        self.action_clear_face.triggered.connect(self.handle_action_clear)

        self.action_next_image = QAction("next_image", self)
        self.action_next_image.triggered.connect(self.handle_action_next_image)

        self.action_previous_image = QAction("previous_image", self)
        self.action_previous_image.triggered.connect(self.handle_action_previous_image)

        self.action_previous_face = QAction("previous_face", self)
        self.action_previous_face.triggered.connect(self.handle_action_previous_face)

        self.action_next_face = QAction("next_face", self)
        self.action_next_face.triggered.connect(self.handle_action_next_face)

        self.action_show_work_queue = QAction("show work queue", self)
        self.action_show_work_queue.triggered.connect(self.handle_show_work_queue)

    def create_imagelist(self):
        image_list = QListView()
        image_list.setEditTriggers(QAbstractItemView.NoEditTriggers)

        self.image_list_model = QStandardItemModel(image_list)
        image_list.setModel(self.image_list_model)

        def on_image_selected():
            if self.image_list_model.rowCount() == 0:
                return

            item = self.get_selected_item()
            if item is not None:
                self.set_selected_slide(item.data(5))

        image_list.setMaximumWidth(200)
        image_list.selectionModel().selectionChanged.connect(on_image_selected)
        self.image_list = image_list

    def set_selected_slide(self, slide):
        if slide is None:
            img = create_placeholder_image((240, 320))
            self.selected_image.setPixmap(image_to_pixmap(img))
            return

        img = cv2.imread(slide.path)
        if img is None:
            img = create_placeholder_image((240, 320))

        self.current_image_faces = [face_to_rect(face) for face in face_recognition.face_locations(img)]

        for face in self.current_image_faces:
            cv2.rectangle(img, face.pt1.to_tuple(), face.pt2.to_tuple(), (200, 200, 200))

        if slide.face_rect is not None:
            pt1, pt2 = slide.face_rect.pt1, slide.face_rect.pt2
            cv2.rectangle(img, (pt1.x, pt1.y), (pt2.x, pt2.y), (0, 255, 0))

        pixmap = image_to_pixmap(img)

        item = self.get_selected_item()
        if item is not None:
            item.setData(slide, 5)

        self.selected_image.setPixmap(pixmap)

    def clear_selected_face(self):
        item = self.get_selected_item()
        if item is None:
            return
        slide = item.data(5)
        self.set_selected_slide(project.InputSlide(slide.path, None))

    def get_selected_item(self):
        indexes = self.image_list.selectedIndexes()
        if len(indexes) == 0:
            return None
        return self.image_list_model.itemFromIndex(indexes[0])

    def handle_action_clear(self):
        self.clear_selected_face()

    def handle_action_next_image(self):
        selected = self.image_list.selectedIndexes()
        if len(selected) == 0:
            self.image_list.setCurrentIndex(self.image_list_model.index(0, 0))
        else:
            row, col = (selected[0].row() + 1) % self.image_list_model.rowCount(), selected[0].column()
            self.image_list.setCurrentIndex(self.image_list_model.index(row, col))

    def handle_action_previous_image(self):
        selected = self.image_list.selectedIndexes()
        list_size = self.image_list_model.rowCount()
        if len(selected) == 0:
            self.image_list.setCurrentIndex(self.image_list_model.index(list_size - 1, 0))
        else:
            row, col = (selected[0].row() + list_size - 1) % self.image_list_model.rowCount(), selected[0].column()
            self.image_list.setCurrentIndex(self.image_list_model.index(row, col))

    def handle_action_previous_face(self):
        self.select_face(-1)

    def handle_action_next_face(self):
        self.select_face(1)

    def handle_show_work_queue(self):
        self.show_work_queue_window()

    def show_work_queue_window(self):
        self.work_queue_window.show()
        self.work_queue_window.activateWindow()
        self.work_queue_window.raise_()

    def select_face(self, offset):
        item = self.get_selected_item()
        if item is None:
            return  # no image selected
        slide = item.data(5)
        if slide is None:
            return  # no image selected

        if len(self.current_image_faces) == 0:
            return  # no faces in current image

        if slide.face_rect is None or slide.face_rect not in self.current_image_faces:
            self.set_selected_slide(project.InputSlide(slide.path, self.current_image_faces[0]))
            return

        current_face_index = self.current_image_faces.index(slide.face_rect)
        if current_face_index < 0:
            self.set_selected_slide(project.InputSlide(slide.path, self.current_image_faces[0]))
        else:
            face = self.current_image_faces[
                (current_face_index + offset + len(self.current_image_faces)) % len(self.current_image_faces)]
            self.set_selected_slide(project.InputSlide(slide.path, face))

    def create_main_widget(self):
        right_panel = QWidget()
        right_panel_layout = QVBoxLayout()
        right_panel_layout.setAlignment(Qt.AlignTop)
        slide_tool_bar = QToolBar()

        def set_face_from_mouseclick(a):
            item = self.get_selected_item()
            if item is None:
                return  # no image selected
            slide = item.data(5)
            if slide is None:
                return  # data for selected image
            face_rect = project.Rect(project.Point(a.x() - 5, a.y() - 5), project.Point(a.x() + 5, a.y() + 5))
            self.set_selected_slide(project.InputSlide(slide.path, face_rect))

        slide_tool_bar.addAction(self.action_previous_image)
        slide_tool_bar.addAction(self.action_next_image)
        slide_tool_bar.addAction(self.action_clear_face)
        slide_tool_bar.addAction(self.action_previous_face)
        slide_tool_bar.addAction(self.action_next_face)

        right_panel_layout.addWidget(slide_tool_bar)
        self.selected_image.mousePressEvent = set_face_from_mouseclick

        right_panel_layout.addWidget(self.selected_image)
        right_panel.setLayout(right_panel_layout)

        main_widget = QWidget()
        main_layout = QHBoxLayout()
        main_layout.setAlignment(Qt.AlignVCenter)
        main_layout.addWidget(self.image_list)
        main_layout.addWidget(right_panel)
        main_widget.setLayout(main_layout)
        self.setCentralWidget(main_widget)

    def add_image(self, path):
        img = cv2.imread(path)
        if img is None:
            return
        faces = face_recognition.face_locations(img)
        if len(faces) == 0:
            self.add_slide(project.InputSlide(path, None))
        else:
            face = faces[0]
            self.add_slide(
                project.InputSlide(path,
                                   project.Rect(project.Point(face[1], face[0]), project.Point(face[3], face[2]))))

    @staticmethod
    def calculate_relative_image_path(image_path, project_path):
        if project_path is None:
            # TODO: given no project path, fall back to cwd?
            return image_path

        project_dir = os.path.dirname(project_path)
        if image_path.startswith(project_dir + '/'):  # TODO: deal with path separator differences across OS
            return image_path[len(project_dir) + 1:]
        return image_path

    def add_slide(self, slide):
        shortened_path = self.calculate_relative_image_path(slide.path, self.current_project_path)
        item = QStandardItem(shortened_path)

        item.setData(slide, 5)
        self.image_list_model.appendRow(item)

    def handle_action_open(self):
        file = QFileDialog.getOpenFileName(self, "Select project", filter="*.json")
        if file == ('', ''):
            return
        self.load_project(file[0])

    def handle_action_new(self):
        self.start_new_project()

    def handle_action_add_image_directory(self):
        file = str(QFileDialog.getExistingDirectory(self, "Select directory"))
        if len(file) == 0:
            return
        self.add_images_from_dir(file)

    def handle_action_save_project(self):
        if self.current_project_path is None:
            file = QFileDialog.getSaveFileName(self, "Save project")[0]
            if file == "":
                return
            self.save_project(file)
        else:
            self.save_project(self.current_project_path)

    def process_next_work_queue_element(self):
        if self.generator_thread is not None:
            print("Generator thread still working")
            return

        work_item = None
        try:
            work_item = next(filter(lambda item: item.status == 'QUEUED', self.work_queue))
        except StopIteration as ex:
            pass

        if work_item is None:
            print("Queue empty")
            return

        def update_work_item(*args):
            state = args[0]
            if state == 'finished':
                work_item.update('FINISHED')
                self.work_queue_window.update()
            else:
                params = args[1:]
                work_item.update('PROCESSING', params[0] / params[1] * 100)
                self.work_queue_window.update()

        def do_generate(proj):
            try:
                generate_facemovie(proj, progress_callback=update_work_item)
            except Exception as ex:
                work_item.update('FAILED', None)
                print("Facemovie generation failed", ex)
            self.generator_thread = None
            self.process_next_work_queue_element()

        self.generator_thread = threading.Thread(target=do_generate, args=[work_item.project])
        self.generator_thread.start()
        self.show_work_queue_window()

    def handle_action_generate_facemovie(self):
        def on_generate_dialog_confirmed(settings):
            self.settings = settings
            self.work_queue.append(WorkItem(self.create_project_from_state()))
            self.process_next_work_queue_element()

        dialog = OutputSettingsWindow(settings=self.settings, generate_callback=on_generate_dialog_confirmed)
        dialog.setAttribute(Qt.WA_DeleteOnClose)
        dialog.exec_()

    def create_toolbar(self):
        toolbar = QToolBar(self)
        toolbar.addAction(self.action_new)
        toolbar.addAction(self.action_open)
        toolbar.addAction(self.action_save)
        toolbar.addAction(self.action_add_image_directory)
        toolbar.addAction(self.action_generate_facemovie)
        toolbar.addAction(self.action_show_work_queue)
        return toolbar

    def save_project(self, file):
        proj = self.create_project_from_state()
        proj.save(file)
        self.save_last_opened(file)

    def create_project_from_state(self):
        proj = project.Project()
        for i in range(self.image_list_model.rowCount()):
            item = self.image_list_model.item(i)
            slide = item.data(5)
            if slide is None:
                continue
            proj.add_slide(project.InputSlide(slide.path, slide.face_rect))
        proj.settings = self.settings
        return proj

    def save_last_opened(self, path):
        rc = self.load_rcfile()
        rc['last_opened'] = path
        self.write_rcfile(rc)

    @staticmethod
    def load_rcfile():
        expanded = os.path.expanduser("~/.facemovierc")
        rc = None
        if os.path.exists(expanded):
            try:
                rc = yaml.load(open(expanded, "r"), Loader=yaml.SafeLoader)
            except Exception:
                print("could not read ~/.facemovierc")
        return dict() if rc is None else rc

    @staticmethod
    def write_rcfile(rc):
        expanded = os.path.expanduser("~/.facemovierc")
        try:
            with open(expanded, "w") as rcfile:
                yaml.dump(rc, rcfile)
        except Exception as e:
            print("could not write ~/.facemovierc")
            print(e)

    def load_project(self, project_file):
        self.start_new_project()
        proj = project.Project.load(project_file)

        self.settings = proj.settings

        self.current_project_path = project_file
        self.save_last_opened(project_file)
        for slide in proj.slides:
            self.add_slide(slide)

    def add_images_from_dir(self, dirname):
        for path in glob.glob(dirname + "/**", recursive=True):
            if not os.path.isfile(path):
                continue
            self.add_image(path)

    def start_new_project(self):
        self.save_last_opened(None)
        self.settings = Settings()
        self.image_list_model.clear()
        self.current_image_faces = []
        self.set_selected_slide(None)


def image_to_pixmap(img):
    return QPixmap(QImage(img, img.shape[1], img.shape[0], img.shape[1] * 3, QImage.Format_RGB888).rgbSwapped())


def create_placeholder_image(shape):
    h, w = shape
    img = np.zeros([h, w, 3], dtype=np.uint8)
    img[:] = 128
    cv2.rectangle(img, (10, 10), (310, 230), (255, 255, 255), 5)
    return img


def main():
    app = QApplication(sys.argv)
    win = MainWindow()
    win.show()
    app.exec()


if __name__ == "__main__":
    main()


gui/components/output_settings/__init__.py
from PyQt5.QtGui import QIntValidator
from PyQt5.QtWidgets import QDialog, QLineEdit, QPushButton, QFileDialog, QGridLayout, QLabel, QComboBox

from gui.constants import RESOLUTIONS
from project import Settings


class OutputSettingsWindow(QDialog):
    def __init__(self, parent=None, settings=None, generate_callback=None):
        super().__init__(parent)
        print("created output settings dialog with settings", settings.fps)
        self.generate_callback = generate_callback

        def create_int_edit(min_value, max_value, initial_value):
            edit = QLineEdit()
            validator = QIntValidator()
            validator.setRange(min_value, max_value)
            edit.setValidator(validator)
            edit.setText(str(initial_value))
            return edit

        def create_file_input():
            edit = QLineEdit()
            return edit

        edit_fade_in_millis = create_int_edit(0, 10000, settings.fade_in_millis)
        edit_show_millis = create_int_edit(0, 10000, settings.show_millis)
        combo_resolution = self.create_resolutions_combobox(RESOLUTIONS)
        edit_fps = create_int_edit(12, 120, settings.fps)
        edit_filename = create_file_input()
        edit_filename.setText(settings.output_path)
        btn_select_output_file = QPushButton("Select file")

        def generate_clicked():
            if self.generate_callback is not None:
                resolution = RESOLUTIONS[combo_resolution.currentIndex()]

                settings = Settings(int(edit_fade_in_millis.text()), int(edit_show_millis.text()), 1000,
                                    (resolution[2], resolution[1]),
                                    int(edit_fps.text()),
                                    edit_filename.text())
                self.generate_callback(settings)
            self.close()

        def select_file_clicked():
            file = QFileDialog.getSaveFileName(self, "Select output filename", initialFilter="*.mov")[0]
            if file == "":
                return
            edit_filename.setText(file)

        btn_select_output_file.clicked.connect(select_file_clicked)

        btn_generate = QPushButton("Generate facemovie")
        btn_generate.clicked.connect(generate_clicked)

        layout = QGridLayout()
        layout.addWidget(QLabel("fade_in_millis"), 0, 0)
        layout.addWidget(edit_fade_in_millis, 0, 1)
        layout.addWidget(QLabel("show_millis"), 1, 0)
        layout.addWidget(edit_show_millis, 1, 1)
        # layout.addWidget(QLabel("fade_out_millis"), 2, 0)
        # layout.addWidget(create_int_edit(0, 10000), 2, 1)
        layout.addWidget(QLabel("target_size"), 2, 0)
        layout.addWidget(combo_resolution, 2, 1)
        layout.addWidget(QLabel("fps"), 3, 0)
        layout.addWidget(edit_fps, 3, 1)
        layout.addWidget(QLabel("Output filename"), 4, 0)
        layout.addWidget(edit_filename, 4, 1)
        layout.addWidget(btn_select_output_file, 4, 2)
        layout.addWidget(btn_generate, 5, 0)

        self.setLayout(layout)

    def create_resolutions_combobox(self, resolutions):
        combo_resolution = QComboBox()
        for i, r in enumerate(resolutions):
            combo_resolution.addItem("%s (%s x %s)" % r)
        return combo_resolution

project.py
import json


class InputSlide:
    def __init__(self, path, face_rect):
        self.path = path
        self.face_rect = face_rect

    def __str__(self):
        return "InputSlide(%s, %s)" % (self.path, self.face_rect)


class Rect:
    def __init__(self, pt1, pt2):
        self.pt1 = pt1
        self.pt2 = pt2

    def to_coordinates(self):
        return [self.pt1.x, self.pt1.y, self.pt2.x, self.pt2.y]

    def __str__(self):
        return "Rect[ %s -> %s ]" % (self.pt1, self.pt2)

    def __eq__(self, other):
        return self.pt1 == other.pt1 and self.pt2 == other.pt2


class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        return "Point[%s,%s]" % (self.x, self.y)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def to_tuple(self):
        return (self.x, self.y)


class Project:
    def __init__(self, playlist=None, slides=None, settings=None):
        if playlist is None:
            playlist = []
        if slides is None:
            slides = []
        if settings is None:
            settings = Settings()

        self.playlist = playlist
        self.slides = slides
        self.settings = settings

    def add_slide(self, slide):
        self.playlist.append((slide.path, None if slide.face_rect is None else slide.face_rect.to_coordinates()))
        self.slides.append(slide)

    def save(self, path):
        data = dict()
        data["playlist"] = self.playlist
        data["settings"] = dict(
            fade_in_millis=self.settings.fade_in_millis,
            fade_out_millis=self.settings.fade_out_millis,
            target_size=self.settings.target_size,
            fps=self.settings.fps,
            output_path=self.settings.output_path,
            show_millis=self.settings.show_millis,
        )
        if path is None:
            print(json.dumps(data))
        else:
            with open(path, 'w') as outfile:
                outfile.write(json.dumps(data))

    @staticmethod
    def load(path):
        with open(path) as infile:
            data = json.load(infile)
            playlist = data["playlist"]
            settings = load_settings(data.get("settings", None))
            return Project(playlist=playlist, slides=[create_slide(item[0], item[1]) for item in playlist],
                           settings=settings)


def load_settings(data=None):
    if data is None:
        return Settings()

    fade_in_millis = data.get("fade_in_millis", 500)
    show_millis = data.get("show_millis", 2000)
    fade_out_millis = data.get("fade_out_millis", 1000)
    target_size = tuple(data.get("target_size", [768, 1024]))
    fps = data.get("fps", 30)
    output_path = data.get("output_path", "out.mov")

    return Settings(
        fade_in_millis=fade_in_millis,
        show_millis=show_millis,
        fade_out_millis=fade_out_millis,
        target_size=target_size,
        fps=fps,
        output_path=output_path
    )


def create_slide(path, face_coordinates):
    rect = create_rect(face_coordinates)
    return InputSlide(path, rect)


def create_rect(coordinates):
    if coordinates is None or len(coordinates) != 4:
        return None
    point1 = Point(coordinates[0], coordinates[1])
    point2 = Point(coordinates[2], coordinates[3])
    return Rect(point1, point2)


class Settings:
    def __init__(self, fade_in_millis=500, show_millis=2000, fade_out_millis=1000, target_size=(768, 1024),
                 fps=25, output_path="out.mov"):
        self.fade_in_millis = fade_in_millis
        self.show_millis = show_millis
        self.fade_out_millis = fade_out_millis
        self.target_size = target_size
        self.fps = fps
        self.output_path = output_path
